corehq/apps/hqadmin/tasks.py

```diff
@@ -201,3 +205,84 @@ def track_pg_limits():
                 cursor.execute(f'select last_value from "{sequence}"')
                 current_value = cursor.fetchone()[0]
                 metrics_gauge('commcare.postgres.sequence.current_value', current_value, {'table': table, 'database': db})
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_cases():
+    _reconcile_es_data('case', 'commcare.elasticsearch.stale_cases', 'reconcile_es_cases')
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="4"))
+def reconcile_es_forms():
+    _reconcile_es_data('form', 'commcare.elasticsearch.stale_forms', 'reconcile_es_forms')
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="6"))
+def count_es_forms_past_window():
+    today = date.today()
+    two_days_ago = today - timedelta(days=2)
+    four_days_ago = today - timedelta(days=4)
+    start = two_days_ago.isoformat()
+    end = four_days_ago.isoformat()
+    _reconcile_es_data(
+        'form',
+        'commcare.elasticsearch.stale_forms_past_window',
+        'es_forms_past_window',
+        start=start,
+        end=end,
+        publish=False
+    )
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="6"))
+def count_es_cases_past_window():
+    today = date.today()
+    two_days_ago = today - timedelta(days=2)
+    four_days_ago = today - timedelta(days=4)
+    start = two_days_ago.isoformat()
+    end = four_days_ago.isoformat()
+    _reconcile_es_data(
+        'case',
+        'commcare.elasticsearch.stale_cases_past_window',
+        'es_cases_past_window',
+        start=start,
+        end=end,
+        publish=False
+    )
+
+
+def _reconcile_es_data(data_type, metric, blob_parent_id, start=None, end=None, republish=True):
+    if not start:
+        today = date.today()
+        two_days_ago = today - timedelta(days=2)
+        start = two_days_ago.isoformat()
+    with NamedTemporaryFile('w', delete=False) as output_file:
+        file_name = output_file.name
+        call_command('stale_data_in_es', data_type, start=start, end=end, stdout=output_file)
```

Comment:
`stale_data_in_es` (for SQL backend) uses a resumable iteration that orders results by primary key. Have you profiled the SQL query to see if it will scan every record in the db or if it is able to use a date index?

I'm having a hard time seeing how it will do an efficient query over a date range with resumable iteration logic that depends on results being ordered by primary key. To do that I think it would need to query all records in the date range before ordering them (which would be inefficient). An alternative that seems more likely is that it will scan all records in a primary key index (every form or case in the system) and filter out all the ones that fall outside the date range.

Reply:
The answer to this question depends a lot on the internal statistics that postgres has on the table, and as such will vary node by node. It is certainly always able to use an index, but using `EXPLAIN` for queries similar to the one in this dbaccessor, it used different ones on different machines (sometimes preferring each of the two methods you described, a pk index and then filtering results on date, or a date index and then sorting by pk). In each case, runtime was comparable, and nothing involved full table scans.